File: Cache_System_Design/Q2. LFU Cache/Q2_LFUCache.py
from collections import defaultdict, OrderedDict

class LFUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity              # Maximum size of cache
        self.minFreq = 0                      # Lowest frequency in cache
        self.keyToNode = {}                   # key -> [value, freq]
        self.freqToList = defaultdict(OrderedDict)  # freq -> OrderedDict of keys

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key not in self.keyToNode:         # Key not found
            return -1

        value, freq = self.keyToNode[key]     # Get value and current freq
        del self.freqToList[freq][key]        # Remove from old freq list

        if not self.freqToList[freq]:         # If old freq list empty
            del self.freqToList[freq]
            if self.minFreq == freq:          # Update minFreq if needed
                self.minFreq += 1

        newFreq = freq + 1                    # Increase frequency
        self.freqToList[newFreq][key] = True  # Add to new freq list
        self.keyToNode[key] = [value, newFreq]  # Update freq

        return value

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: None
        """
        if self.capacity == 0:                # If no space, do nothing
            return

        if key in self.keyToNode:             # If key already exists
            self.keyToNode[key][0] = value    # Update value
            self.get(key)                     # Reuse get() to update freq
            return

        if len(self.keyToNode) == self.capacity:   # Cache full → evict LFU
            freq_list = self.freqToList[self.minFreq]
            evict_key, _ = freq_list.popitem(last=False)  # Remove LRU of LFU freq
            if not freq_list:
                del self.freqToList[self.minFreq]
            del self.keyToNode[evict_key]

        # Insert new key with freq=1
        self.keyToNode[key] = [value, 1]
        self.freqToList[1][key] = True
        self.minFreq = 1                      # Reset minimum frequency

File: Cache_System_Design/Q2. LFU Cache/Q2_LFUCache.java
import java.util.*;

class LFUCache {

    // Node to store key, value, and frequency
    private class Node {
        int key, value, freq;
        Node(int key, int value) {
            this.key = key;       // store key
            this.value = value;   // store value
            this.freq = 1;        // new key always starts with freq = 1
        }
    }

    private int capacity;  // max size of cache
    private int minFreq;   // smallest frequency in cache
    private Map<Integer, Node> keyToNode; // map key -> node
    private Map<Integer, LinkedHashSet<Integer>> freqToList; // map freq -> keys with that freq

    public LFUCache(int capacity) {
        this.capacity = capacity;         // set capacity
        this.minFreq = 0;                 // no freq yet
        keyToNode = new HashMap<>();      // store key -> node
        freqToList = new HashMap<>();     // store freq -> set of keys
    }

    public int get(int key) {
        if (!keyToNode.containsKey(key))  // if key not found
            return -1;                    // return -1

        Node node = keyToNode.get(key);   // get the node
        updateFrequency(node);            // increase frequency
        return node.value;                // return value
    }

    public void put(int key, int value) {
        if (capacity == 0) return;        // if size 0, do nothing

        if (keyToNode.containsKey(key)) { // if key already exists
            Node node = keyToNode.get(key); 
            node.value = value;           // update value
            updateFrequency(node);        // update frequency
            return;                       // done
        }

        if (keyToNode.size() == capacity) { // cache full
            LinkedHashSet<Integer> list = freqToList.get(minFreq); // get keys with smallest freq
            int evictKey = list.iterator().next(); // first inserted = LRU
            list.remove(evictKey);       // remove from freq list
            if (list.isEmpty())          // if no keys left in this freq
                freqToList.remove(minFreq); // remove freq entry
            keyToNode.remove(evictKey);  // remove key from map
        }

        Node newNode = new Node(key, value); // create new key
        keyToNode.put(key, newNode);         // store node in map

        freqToList.computeIfAbsent(1, k -> new LinkedHashSet<>()) // get list for freq=1
                    .add(key);               // add new key

        minFreq = 1;                          // reset minimum frequency to 1
    }

    private void updateFrequency(Node node) {
        int oldFreq = node.freq;             // get old frequency
        int newFreq = oldFreq + 1;           // new frequency = old + 1

        LinkedHashSet<Integer> oldList = freqToList.get(oldFreq); // get old freq list
        oldList.remove(node.key);           // remove key from old freq
        if (oldList.isEmpty()) {            // if list empty
            freqToList.remove(oldFreq);     // remove freq entry
            if (minFreq == oldFreq)         // if this was the minimum freq
                minFreq++;                  // increase minimum freq
        }

        freqToList.computeIfAbsent(newFreq, x -> new LinkedHashSet<>()) // get new freq list
                  .add(node.key);           // add key to new freq

        node.freq = newFreq;                // update node's freq
    }
}
